### ISSUE TITLE
Make output_type public in the output data structure

### ISSUE THREAD

**Issue by pdobacz** [omisego/plasma-contracts#290]
Following our discussion on a call, I'm suggesting that the `output_type`, now planned to be included in the obfuscated data structure (`output_guard`), is made public, similar to how `tx_type` is handled.

Below is the description of the issue and suggested change to the ALD.

Let's take some time to think about the implications.

## Issue Type 

```
[ ] bug report
[x] feature request (design)
```

## Current Behavior

The current design keeps the `output_type` included in the `output_guard_preimage` and is delivered in the transaction data obfuscated. This means:
 - we're conflating two distinct mechanism - the exchange deposit obfuscation mechanism and the ALD's `output_type` discovery
 - we're limiting the discovery of `output_type` only if you have the remainder of the preimage (think: a query to get all outputs of certain type)
 - we're making the discovery of the `owner` of outputs harder (think: a query to get all outputs of all types belonging to someone - you need to know all the possible types)
 - we're making modifications to the decoding of an output harder - one needs the witness data (preimage) to discover the output type (think: changing the encoding of an output entirely)
 - we're making the so-called 1:1 binding more difficult: since `output_type` determines `input_pointer_type`, child chain/watcher wouldn't know how to store the output in their UTXO set. (1:1 binding makes UTXOset management easier and more abstract and also double-spend detection easier, see [here](https://github.com/omisego/research/issues/93#issuecomment-517734509) for more info)
 - we're forcing ourselves to include things like `confirmer_address` in the `output_guard_preimage` too, thereby obfuscating it and leading to similar problems as written above

## Expected Behavior

`output_type` is delivered similar to `tx_type` as the first n bit field of the structure (presumably RLP encodable array), in public. The obfuscation mechanism is set separate from the `output_type` typing - `output_guard` may contain either the hashed preimage or (hashed) owner's address, **keeping the `output_type` the same**. Then the witness to unlock such output may contain either the preimage (+other data if applicable) or the owners signature.

## Motivation for Change

If the `output_type` is public:
 - we can dispatch do different decoding schemes if the current one turns out to be insufficient. An example of this is that we can then include `confirm_address` of an MVP output in the public section of the output data structure
 - we're keeping the two mechanisms (dispatch & obfuscation) separate
 - we still can have >1 `output_type` for a single creating `tx_type`, which seems to be required
 - we have the same dispatch/typing mechanism for output and transaction

**Comment by paulperegud** [comment]
LGTM

**Comment by paulperegud** [comment]
@boolafish what do you think about this one?

**Comment by boolafish** [comment]
> we can dispatch do different decoding schemes if the current one turns out to be insufficient. 

Some implementation detail concerns. If a tx type has different output type with different decode schema, we need something more abstract on our current rlp encoding to output. The output itself (probably) need to be double encoded to be able to allow an array of arbitrary type (`bytes`) and decode on demand. 

Otherwise, if we are using custom `struct` to be the type of the outputs array, we cannot have one array with two `structs` then we'd need to have multiple arrays and then it becomes a problem how we map the `outputType` from the array to the index in which array.

```
tx: {
    txType: uint256,
    outputTypes: [uint256, ...],
    inputs: [bytes32, ...],
    outputs: [bytes, ...],  <-- need to be same "type" for all outputs, so need to rlp encode the output itself?
    other fields....
}
```

So question, do we want to change our output encoding schema now? Quite likely some efforts on elixir side so I am not sure how heavy would that be. (btw, let's discuss this with #282 on what scope we'd *like* to do also according to our timeline, what *can* we do)

----

Some breaking changes with current smart contract codes and with previous design for DEX (that I've sensed)

1. One output type that can spent in both `PaymentTx` and `DexSettlement`. Previously, it was designed to be different output type.
1. We need a way to get `confirm sig address`. We lock in the code of using `output guard` to get confirm sig address now, need to change that to get from `output`. (I am thinking replacing existing `OutputGuardHandler` to become sth like `OutputHandler` instead, so it can do things via output guard or output as output guard would be part of output anyway.
1. How do we get `exitTarget` (the address to exit/withdraw fund to). Currently, we do it via `OutputGuardHandler` which each `outputType` has its implementation. However, if we are moving to one output type handling both scenario, where do we put the flag to differentiate? In the pre-image of output guard? Then do we hash the output guard for normal payment output as well?

**Comment by pdobacz** [comment]
Thanks! ooph, this looks like a substantial effort with the current impl :/. My gut tells me it's unlikely to go into the current immediate scope.

On top of this, just a couple of points I can relate to:

> Quite likely some efforts on elixir side so I am not sure how heavy would that be

Abstract output handling hasn't yet landed on elixir side (actually, it would be good to see where this discussion is headed for a finalized version of this elixir impl.). Also, the current attempt to implement it there (wrongly) assumed that the output type is public, becaue of my incomplete knowledge, so it's probably actually less work with public `output_type` in `elixir-omg`. BTW, my wrong assumption was kinda done by extrapolating the `tx_type` mechanism.

> breaking changes with (...) previous design for DEX
> ...
> One output type that can spent in both PaymentTx and DexSettlement. Previously, it was designed to be different output type.

It would be good to double check against the DEX specs, but I think it's OK. I.e. it all boils down to whether the spending condition says :+1: or :-1:, prividing an API layer in between DEX protocol and this. But yea, revisiting would be necessary to make sure this holds water.

**Comment by boolafish** [comment]
> It would be good to double check against the DEX specs, but I think it's OK.

It need to be same type otherwise it would leak that the trader is doing DEX deposit.

**Comment by pdobacz** [comment]
Notes after the discussion w/ @kevsul @boolafish @paulperegud @Pongch :

**tl;dr** now only introduce `output_type`, punt the hard implementation stuff till new ExitGame (a future upgrade)

Implementation hurdles towards doing a full-fledged abstract output + output_type public in the contract:
 - opaque types missing in solidity (essentially what Andy posted above).
     - one solution: handling RLP encoded blobs instead of structs and decoding as late as possible sounds feasible and should not be too much gas.
     - alternative solution: same output structure for all `output_types`:
        ```
        output_type
        output_guard - ....
        token_standard <- new field, uint
        token_contract_address - address
        standard_specific_data - bytes
        output_type_specific_data - bytes
        ```
        This is a variation of the RLP encoded blobs - the blob (`output_type_specific_data`) sits as a field in a generic struct and is decoded only when needed (decided based on `output_type`) . This solves the confirm sig address being or not being there, still retaining the ability to define a solidity struct to hold the `Output`.
  - discovering exit target would require witness/preimage, but it does anyway so no biggie

Doing the full implementation along the lines of :point_up: seems unfeasible **BUT**:

`output_type` can be made public and put in the first item of the RLP encoded output structure.
**The rest** would stay the same, i.e. the structure would be fixed and we rely on the upgrade to a new `ExitGame` to provide full output type abstractness, possibly as outlined above.

We are thus able to make a cheap modification, that will provide the minimal future-proofing required and allow introduction of abstract output types later on.
It seems that the upgrade mechanism (new ExitGame) will be OK with processing an entirely new output data structure.

Kudos to @kevsul @boolafish @paulperegud  for realizing this.

----------------

There is one pending detail to decide on. There's 2 alternatives on passing `output_type` around:
  - as a first entry in an RLP encoded data for every output instance, i.e. `encoded_output=RLP.encode(output_type, other_stuffs)`
  - as a list in the second entry in the RLP encoded data for _every tx instance_, i.e. `encoded_output=RLP.encode(other_stuffs)`, `encoded_tx=RLP.encode(tx_type, [output_type1, output_type2, ...], other_stuffs)`

**Comment by pdobacz** [comment]
When this is implemented, let's not forget to update RC MVP DEX and ALD specs that we might have fleshed out by then, which are, as of now, outside this repo.

**Comment by kevsul** [comment]
> There is one pending detail to decide on. There's 2 alternatives on passing output_type around:
> - as a first entry in an RLP encoded data for every output instance, i.e. encoded_output=RLP.encode(output_type, other_stuffs)
> - as a list in the second entry in the RLP encoded data for every tx instance, i.e. encoded_output=RLP.encode(other_stuffs), encoded_tx=RLP.encode(tx_type, [output_type1, output_type2, ...], other_stuffs)

The first option feels more natural to me, keeping the output type with the rest of the output data.

@boolafish was there a particular reason you proposed the second option?

**Comment by boolafish** [comment]
@kevsul no really strong reason as I re-think, probably just my brain not working really well that time XD

Pros:
1. slightly more flexible, less assumption required. Assuming `type` to always be some first positions of `tx` make the remaining position of tx free of abstraction if we ever need that.
1. Potentially code wise it is more compatible with our existing one as we separate `outputType` from `output` now in everyplace.

Cons:
`outputType` is further away from the `output` itself so we always need to treat them kinda separated in a way....

**Comment by pdobacz** [comment]
to me, option (1.) has two benefits, expanding on what was stated above:
 - same typing mechanism for `tx` and `output`
 - `output_type` is still accessible, when we're processing an `output` instance, but have no access to the `tx` instance it came from; `output` data structure is self-contained

A downside of (1.) I can think of is this:
 - if one's really only interested in discovering the types of outputs, given a `tx`, going through outputs and having to `RLP.decode()` all of them just to fetch the first item is more gas-costly. But solving this particular problem is optimizable, if we're desperate to cut gas costs (`RLP.decode_only_first_item()`)

**Comment by kevsul** [comment]
> - output_type is still accessible, when we're processing an output instance, but have no access to the tx instance it came from; output data structure is self-contained

Yeah, encapsulating the outputType in the output is key for me. I'm strongly leaning towards this option - unless I come across anything that's clearly better the other way, I'm going to go with this one.
